rfcs/len-metamethod-rawlen.md

```diff
@@ -20,7 +20,7 @@ Supporting `__len` would make it possible to implement a custom integer based co
 
 ## Design
 
-`#v` will call `__len` metamethod if the object is a table and the metamethod exists; the result of the metamethod will be returned.
+`#v` will call `__len` metamethod if the object is a table and the metamethod exists; the result of the metamethod will be returned if it's a number (an error will be raised otherwise).
```

Comment:
This seems a bit inconsistent with other metamethods.

`__concat`, `__unm`, and all the binary/math operator metamethods, don't seem to do any checks.
`__eq`, `__le` and `__lt` convert the return to a boolean silently, seemingly via truthiness rules.

On the other side of the argument, `__tostring` _does_ enforce a string return.

Reply:
I don't think `__unm`/`__concat` are comparable:

`__unm` doesn't have to return a number - in fact, it should likely return the value of the same type as its argument for arithmetic consistency.
`__concat` is likely to be used in contexts where "concatenation" similarly should preserve the source type instead of manufacturing a string; for example, container types may override `..` to mean "container concatenation", eg returning a new container that contains elements from both sides.

I agree with `eq` etc, they should ideally be restricted to booleans but it's not strictly a necessary change (since every type is convertable to a boolean implicitly anyway), so there's already a guarantee that a caller has that `a == b` always evaluates to a boolean.

However, not every type can be converted to a number, so `#` either has to fail on non-numbers, or return the value of arbitrary type to the caller. The latter is worse from the type safety perspective; since today `#` can only return a number and it's not clear why `#` should be able to return a value of another type, it seems beneficial to keep it as such.

Essentially the general rule would be that if it makes sense semantically to return an arbitrary type (like `__unm`/`__concat`), then we should not restrict the returned type; if there's realistically only one semantically meaningful returned type, then it's better to guarantee to the caller that you either get the result of that type or an error.

Reply:
Would the `#` operation cast the return to unsigned 8/16/32/64/128/etc bit integers (if possible) akin to some standard library functions (e.g. math.random for signed and bit32 for unsigned) or would `#` allow all values from the VM 'number' data type (IEC 559 binary64) so instead of just nonnegative ℤ, it'll include ℝ (actually a subset of ℚ with the denominator of 2^k), ±Infinity, and ±NaNQ(x)/±NaNS(x)/Ind?
I believe that `#` only returns nonnegative integers so it might as will cast it to a nonnegative integer data type.